go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -1402,34 +1390,49 @@ func (tsv *TabletServer) convertAndLogError(ctx context.Context, sql string, bin
 	// 2. FAILED_PRECONDITION errors. These are caused when a failover is in progress.
 	// If so, we don't want to suppress the error. This will allow VTGate to
 	// detect and perform buffering during failovers.
-	if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
-		sqlErr, ok := origErr.(*mysql.SQLError)
-		if ok {
-			sqlState := sqlErr.SQLState()
-			errnum := sqlErr.Number()
-			err = vterrors.Errorf(errCode, "(errno %d) (sqlstate %s) during query: %s", errnum, sqlState, sqlparser.TruncateForLog(sql))
-			err = formatErrorWithCallerID(ctx, err)
+	var message string
+	sqlErr, ok := err.(*mysql.SQLError)
+	if ok {
+		sqlState := sqlErr.SQLState()
+		errnum := sqlErr.Number()
+		if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
+			err = vterrors.Errorf(errCode, "(errno %d) (sqlstate %s)%s during query: %s", errnum, sqlState, callerID, sqlparser.TruncateForLog(sql))
+			message = err.Error()
+		} else {
+			err = vterrors.Errorf(errCode, "%s (errno %d) (sqlstate %s)%s: %s", sqlErr.Message, errnum, sqlState, callerID, queryAsString(sql, bindVariables))
+			message = err.Error()
+		}
+	} else {
+		err = vterrors.Errorf(errCode, "%v%s", err.Error(), callerID)
+		if tsv.TerseErrors && len(bindVariables) != 0 && errCode != vtrpcpb.Code_FAILED_PRECONDITION {
+			message = fmt.Sprintf("%v: %v", err, queryAsString(sql, nil))
+		} else {
+			message = fmt.Sprintf("%v: %v", err, queryAsString(sql, bindVariables))
 		}
 	}
 
+	// In order to correctly truncate long queries in logs, combine
+	// the error (which contains the mysql error string)
+	// with the unexpanded query + bind variables, then
+	// truncate the resulting string.
+	//
+	// This makes it so that the query portion of the logs is
+	// properly truncated to the expected max log length.
+	maxLen := *sqlparser.TruncateErrLen
+	if maxLen != 0 && len(message) > maxLen {
+		message = message[:maxLen-12] + " [TRUNCATED]"
+	}
+	if logMethod != nil {
```

Comment:
I would move this up before line 1414, so we don't truncate the message if we're not going to use it.